### ISSUE TITLE
question: Does class-transformer supports complex Circular structures?

### ISSUE THREAD

**Issue by Judahh** [typestack/class-transformer#605]
**I was trying to...**
use class-transformer on complex Circular structures.
It has mongoose connections instances, circular dependancies, and a map to binded functions used by a pub/sub pattern structure.

**The problem:**
```
mongoose: To create a new ObjectId please try `Mongoose.Types.ObjectId` instead of using `Mongoose.Schema.ObjectId`. Set the `suppressWarning` option if you're trying to create a hex char path in your schema.
Trace: 
    at ObjectId [as type] (/Users/judah/auth/node_modules/mongoose/lib/schema/objectid.js:35:13)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:190:37)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at ClassTransformer.classToPlain (/Users/judah/auth/node_modules/src/ClassTransformer.ts:25:21)
    at ClassTransformer.serialize (/Users/judah/auth/node_modules/src/ClassTransformer.ts:137:32)
    at Object.serialize (/Users/judah/auth/node_modules/src/index.ts:93:27)
    at Object.<anonymous> (/Users/judah/auth/next.config.ts:37:12)
    at Module._compile (internal/modules/cjs/loader.js:1075:30)
    at Module.m._compile (/Users/judah/auth/node_modules/ts-node/src/index.ts:1056:23)
    at Module._extensions..js (internal/modules/cjs/loader.js:1096:10)
    at Object.require.extensions.<computed> [as .ts] (/Users/judah/auth/node_modules/ts-node/src/index.ts:1059:12)
    at Module.load (internal/modules/cjs/loader.js:940:32)
    at Function.Module._load (internal/modules/cjs/loader.js:781:14)
    at Module.require (internal/modules/cjs/loader.js:964:19)
mongoose: To create a new ObjectId please try `Mongoose.Types.ObjectId` instead of using `Mongoose.Schema.ObjectId`. Set the `suppressWarning` option if you're trying to create a hex char path in your schema.
Trace: 
    at ObjectId [as id] (/Users/judah/auth/node_modules/mongoose/lib/schema/objectid.js:35:13)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:190:37)
    at /Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:79:30
    at Array.forEach (<anonymous>)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:44:24)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at ClassTransformer.classToPlain (/Users/judah/auth/node_modules/src/ClassTransformer.ts:25:21)
    at ClassTransformer.serialize (/Users/judah/auth/node_modules/src/ClassTransformer.ts:137:32)
    at Object.serialize (/Users/judah/auth/node_modules/src/index.ts:93:27)
    at Object.<anonymous> (/Users/judah/auth/next.config.ts:37:12)
    at Module._compile (internal/modules/cjs/loader.js:1075:30)
    at Module.m._compile (/Users/judah/auth/node_modules/ts-node/src/index.ts:1056:23)
    at Module._extensions..js (internal/modules/cjs/loader.js:1096:10)
    at Object.require.extensions.<computed> [as .ts] (/Users/judah/auth/node_modules/ts-node/src/index.ts:1059:12)
mongoose: To create a new ObjectId please try `Mongoose.Types.ObjectId` instead of using `Mongoose.Schema.ObjectId`. Set the `suppressWarning` option if you're trying to create a hex char path in your schema.
Trace: 
    at ObjectId [as sponsorId] (/Users/judah/auth/node_modules/mongoose/lib/schema/objectid.js:35:13)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:190:37)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at ClassTransformer.classToPlain (/Users/judah/auth/node_modules/src/ClassTransformer.ts:25:21)
    at ClassTransformer.serialize (/Users/judah/auth/node_modules/src/ClassTransformer.ts:137:32)
    at Object.serialize (/Users/judah/auth/node_modules/src/index.ts:93:27)
    at Object.<anonymous> (/Users/judah/auth/next.config.ts:37:12)
    at Module._compile (internal/modules/cjs/loader.js:1075:30)
    at Module.m._compile (/Users/judah/auth/node_modules/ts-node/src/index.ts:1056:23)
    at Module._extensions..js (internal/modules/cjs/loader.js:1096:10)
    at Object.require.extensions.<computed> [as .ts] (/Users/judah/auth/node_modules/ts-node/src/index.ts:1059:12)
    at Module.load (internal/modules/cjs/loader.js:940:32)
    at Function.Module._load (internal/modules/cjs/loader.js:781:14)
    at Module.require (internal/modules/cjs/loader.js:964:19)
    at require (internal/modules/cjs/helpers.js:88:18)
mongoose: To create a new ObjectId please try `Mongoose.Types.ObjectId` instead of using `Mongoose.Schema.ObjectId`. Set the `suppressWarning` option if you're trying to create a hex char path in your schema.
Trace: 
    at ObjectId [as id] (/Users/judah/auth/node_modules/mongoose/lib/schema/objectid.js:35:13)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:190:37)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at ClassTransformer.classToPlain (/Users/judah/auth/node_modules/src/ClassTransformer.ts:25:21)
    at ClassTransformer.serialize (/Users/judah/auth/node_modules/src/ClassTransformer.ts:137:32)
    at Object.serialize (/Users/judah/auth/node_modules/src/index.ts:93:27)
    at Object.<anonymous> (/Users/judah/auth/next.config.ts:37:12)
    at Module._compile (internal/modules/cjs/loader.js:1075:30)
    at Module.m._compile (/Users/judah/auth/node_modules/ts-node/src/index.ts:1056:23)
    at Module._extensions..js (internal/modules/cjs/loader.js:1096:10)
    at Object.require.extensions.<computed> [as .ts] (/Users/judah/auth/node_modules/ts-node/src/index.ts:1059:12)
    at Module.load (internal/modules/cjs/loader.js:940:32)
    at Function.Module._load (internal/modules/cjs/loader.js:781:14)
    at Module.require (internal/modules/cjs/loader.js:964:19)
    at require (internal/modules/cjs/helpers.js:88:18)
TypeError: Cannot convert undefined or null to object
    at Object.transform (/Users/judah/auth/node_modules/flexiblepersistence/source/database/noSQL/mongoDB/mongoDB.ts:45:9)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:190:37)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at TransformOperationExecutor.transform (/Users/judah/auth/node_modules/src/TransformOperationExecutor.ts:307:31)
    at ClassTransformer.classToPlain (/Users/judah/auth/node_modules/src/ClassTransformer.ts:25:21)
    at ClassTransformer.serialize (/Users/judah/auth/node_modules/src/ClassTransformer.ts:137:32)
    at Object.serialize (/Users/judah/auth/node_modules/src/index.ts:93:27)
    at Object.<anonymous> (/Users/judah/auth/next.config.ts:37:12)
    at Module._compile (internal/modules/cjs/loader.js:1075:30)
    at Module.m._compile (/Users/judah/auth/node_modules/ts-node/src/index.ts:1056:23)
    at Module._extensions..js (internal/modules/cjs/loader.js:1096:10)
    at Object.require.extensions.<computed> [as .ts] (/Users/judah/auth/node_modules/ts-node/src/index.ts:1059:12)
    at Module.load (internal/modules/cjs/loader.js:940:32)
    at Function.Module._load (internal/modules/cjs/loader.js:781:14)
(node:8946) Warning: Accessing non-existent property 'MongoError' of module exports inside circular dependency
```

**Comment by NoNameProvided** [comment]
There is an `enableCircularCheck` option when calling `classToPlain`. However, the support for Mongoose objects is a little bit wobbly. You probably should transform directly your `_id` property with a `@Transform` decorator or exclude it.

**Comment by NoNameProvided** [comment]
This should be fixed in 90feca36d2b44b6c66428ccef9402b13b58f7e2e.
